File: rustspider/src/distributed/redis_distributed.rs
//! 分布式模块（完整版）
//! 支持 Redis 分布式队列、布隆过滤器、任务分发

use md5::{Digest, Md5};
use redis::{Client, Commands};
use serde::{Deserialize, Serialize};
use std::collections::hash_map::DefaultHasher;
use std::error::Error;
use std::hash::{Hash, Hasher};
use std::time::{SystemTime, UNIX_EPOCH};

/// 爬取任务
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlTask {
    pub url: String,
    pub priority: i32,
    pub depth: i32,
    pub task_type: String,
    pub spider_name: String,
    pub created_at: f64,
    pub retry_count: i32,
    pub metadata: Option<serde_json::Value>,
}

impl CrawlTask {
    pub fn new(url: String) -> Self {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        Self {
            url,
            priority: 0,
            depth: 0,
            task_type: "crawl".to_string(),
            spider_name: "default".to_string(),
            created_at: now,
            retry_count: 0,
            metadata: None,
        }
    }

    pub fn with_priority(mut self, priority: i32) -> Self {
        self.priority = priority;
        self
    }

    pub fn with_depth(mut self, depth: i32) -> Self {
        self.depth = depth;
        self
    }

    pub fn with_spider_name(mut self, name: String) -> Self {
        self.spider_name = name;
        self
    }

    pub fn to_json(&self) -> Result<String, Box<dyn Error>> {
        Ok(serde_json::to_string(self)?)
    }

    pub fn from_json(json_str: &str) -> Result<Self, Box<dyn Error>> {
        Ok(serde_json::from_str(json_str)?)
    }
}

/// Redis 分布式队列
pub struct RedisDistributedQueue {
    client: Client,
    name: String,
    max_size: usize,
}

impl RedisDistributedQueue {
    /// 创建队列
    pub fn new(redis_url: &str, name: &str, max_size: usize) -> Result<Self, Box<dyn Error>> {
        let client = Client::open(redis_url)?;
        Ok(Self {
            client,
            name: name.to_string(),
            max_size,
        })
    }

    /// 添加任务
    pub fn push(&self, task: &CrawlTask) -> Result<bool, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;

        // 检查队列大小
        let size: usize = conn.zcard(self.priority_queue_key())?;
        if size >= self.max_size {
            return Ok(false);
        }

        // 检查是否已存在
        if self.exists(&task.url)?
            || conn.sismember::<_, _, bool>(self.pending_key(), &task.url)?
            || conn.hexists::<_, _, bool>(self.processing_key(), &task.url)?
        {
            return Ok(false);
        }

        // 添加到优先级队列
        let task_json = task.to_json()?;
        let _: () = conn.zadd(self.priority_queue_key(), &task_json, task.priority)?;
        let _: () = conn.sadd(self.pending_key(), &task.url)?;

        Ok(true)
    }

    /// 获取任务
    pub fn pop(&self) -> Result<Option<CrawlTask>, Box<dyn Error>> {
        self.lease("scheduler", 30)
    }

    pub fn lease(
        &self,
        worker_id: &str,
        lease_ttl_secs: u64,
    ) -> Result<Option<CrawlTask>, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;

        let results: Vec<(String, f64)> = conn.zpopmax(self.priority_queue_key(), 1)?;

        if let Some((task_json, _score)) = results.into_iter().next() {
            let task = CrawlTask::from_json(&task_json)?;
            let _: () = conn.srem(self.pending_key(), &task.url)?;

            let processing_data = serde_json::json!({
                "url": task.url,
                "started_at": current_timestamp(),
                "worker_id": worker_id,
                "expires_at": current_timestamp() + lease_ttl_secs as f64,
                "retry_count": task.retry_count,
                "task": task,
            });
            let _: () = conn.hset(
                self.processing_key(),
                self.url_key(&task.url),
                processing_data.to_string(),
            )?;

            return Ok(Some(task));
        }

        Ok(None)
    }

    /// 确认任务完成
    pub fn ack(&self, task: &CrawlTask, success: bool) -> Result<(), Box<dyn Error>> {
        self.ack_url(&task.url, success, 3)
    }

    pub fn ack_url(
        &self,
        url: &str,
        success: bool,
        max_retries: i32,
    ) -> Result<(), Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;

        let payload: Option<String> = conn.hget(self.processing_key(), self.url_key(url))?;
        let _: () = conn.hdel(self.processing_key(), self.url_key(url))?;
        let Some(payload) = payload else {
            return Ok(());
        };
        let lease: serde_json::Value = serde_json::from_str(&payload)?;
        let mut task: CrawlTask = serde_json::from_value(lease["task"].clone())?;

        if success {
            let _: () = conn.sadd(self.url_set_key(), self.url_key(url))?;
            return Ok(());
        }

        task.retry_count += 1;
        if task.retry_count > max_retries {
            let _: () = conn.lpush(self.failed_queue_key(), task.to_json()?)?;
            let _: () = conn.sadd(self.url_set_key(), self.url_key(url))?;
        } else {
            let _: () = conn.zadd(self.priority_queue_key(), task.to_json()?, task.priority)?;
            let _: () = conn.sadd(self.pending_key(), url)?;
        }
        Ok(())
    }

    pub fn heartbeat(&self, url: &str, lease_ttl_secs: u64) -> Result<bool, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let key = self.url_key(url);
        let payload: Option<String> = conn.hget(self.processing_key(), &key)?;
        let Some(payload) = payload else {
            return Ok(false);
        };
        let mut lease: serde_json::Value = serde_json::from_str(&payload)?;
        lease["expires_at"] = serde_json::Value::from(current_timestamp() + lease_ttl_secs as f64);
        let _: () = conn.hset(self.processing_key(), key, lease.to_string())?;
        Ok(true)
    }

    pub fn reap_expired_leases(
        &self,
        now_secs: f64,
        max_retries: i32,
    ) -> Result<usize, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let processing: std::collections::HashMap<String, String> =
            conn.hgetall(self.processing_key())?;
        let mut reaped = 0usize;
        for (url_key, payload) in processing {
            let lease: serde_json::Value = serde_json::from_str(&payload)?;
            let expires_at = lease["expires_at"].as_f64().unwrap_or(0.0);
            if expires_at > now_secs {
                continue;
            }
            let task_url = lease["url"].as_str().unwrap_or_default().to_string();
            self.ack_url(&task_url, false, max_retries)?;
            reaped += 1;
        }

        Ok(reaped)
    }

    /// 检查 URL 是否存在
    pub fn exists(&self, url: &str) -> Result<bool, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let exists: bool = conn.sismember(self.url_set_key(), self.url_key(url))?;
        Ok(exists)
    }

    /// 队列大小
    pub fn size(&self) -> Result<usize, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let size: usize = conn.zcard(self.priority_queue_key())?;
        Ok(size)
    }

    /// 处理中任务数
    pub fn processing_count(&self) -> Result<usize, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let count: usize = conn.hlen(self.processing_key())?;
        Ok(count)
    }

    /// 失败队列大小
    pub fn failed_count(&self) -> Result<usize, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let count: usize = conn.llen(self.failed_queue_key())?;
        Ok(count)
    }

    /// 清空队列
    pub fn clear(&self) -> Result<(), Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let _: () = conn.del(&[
            self.priority_queue_key(),
            self.url_set_key(),
            self.processing_key(),
            self.failed_queue_key(),
        ])?;
        Ok(())
    }

    /// 获取统计信息
    pub fn get_stats(&self) -> Result<serde_json::Value, Box<dyn Error>> {
        Ok(serde_json::json!({
            "queue_size": self.size()?,
            "processing": self.processing_count()?,
            "failed": self.failed_count()?,
            "max_size": self.max_size,
        }))
    }

    fn priority_queue_key(&self) -> String {
        "spider:shared:queue".to_string()
    }

    fn url_set_key(&self) -> String {
        "spider:shared:visited".to_string()
    }

    fn processing_key(&self) -> String {
        format!("queue:{}:processing", self.name)
    }

    fn pending_key(&self) -> String {
        format!("queue:{}:pending", self.name)
    }

    fn failed_queue_key(&self) -> String {
        format!("queue:{}:failed", self.name)
    }

    fn url_key(&self, url: &str) -> String {
        let mut hasher = Md5::new();
        hasher.update(url.as_bytes());
        format!("url:{:x}", hasher.finalize())
    }
}

/// Redis 布隆过滤器
pub struct RedisBloomFilter {
    client: Client,
    name: String,
    size: usize,
    hash_count: usize,
}

impl RedisBloomFilter {
    /// 创建布隆过滤器
    pub fn new(
        redis_url: &str,
        name: &str,
        expected_items: usize,
        error_rate: f64,
    ) -> Result<Self, Box<dyn Error>> {
        let client = Client::open(redis_url)?;

        // 计算最优参数
        let size = (((-(expected_items as f64)) * error_rate.ln()) / (2.0_f64.ln().powi(2))).ceil()
            as usize;
        let size = size.div_ceil(8) * 8; // 对齐到字节
        let hash_count = ((size as f64 / expected_items as f64) * 2.0_f64.ln()).ceil() as usize;

        Ok(Self {
            client,
            name: name.to_string(),
            size,
            hash_count,
        })
    }

    /// 添加元素
    pub fn add(&self, item: &str) -> Result<bool, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let item_bytes = item.as_bytes();
        let mut is_new = true;

        for seed in 0..self.hash_count {
            let pos = self.hash(item_bytes, seed) % self.size;
            let existed: bool = conn.setbit(self.key(), pos, true)?;
            if !existed {
                is_new = false;
            }
        }

        Ok(is_new)
    }

    /// 检查元素是否存在
    pub fn contains(&self, item: &str) -> Result<bool, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let item_bytes = item.as_bytes();

        for seed in 0..self.hash_count {
            let pos = self.hash(item_bytes, seed) % self.size;
            let exists: bool = conn.getbit(self.key(), pos)?;
            if !exists {
                return Ok(false);
            }
        }

        Ok(true)
    }

    /// 清空过滤器
    pub fn clear(&self) -> Result<(), Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let _: () = conn.del(self.key())?;
        Ok(())
    }

    /// 估算元素数量
    pub fn count(&self) -> Result<usize, Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;
        let bits_set: usize = conn.bitcount(self.key())?;

        if bits_set == 0 {
            return Ok(0);
        }

        let count = -(self.size as f64 / self.hash_count as f64)
            * (1.0 - bits_set as f64 / self.size as f64).ln();

        Ok(count as usize)
    }

    fn key(&self) -> String {
        format!("bloom:{}", self.name)
    }

    fn hash(&self, item: &[u8], seed: usize) -> usize {
        let mut hasher = DefaultHasher::new();
        item.hash(&mut hasher);
        seed.hash(&mut hasher);
        hasher.finish() as usize
    }
}

/// Redis 分布式调度器
pub struct RedisDistributedScheduler {
    queue: RedisDistributedQueue,
    filter: RedisBloomFilter,
    spider_name: String,
    client: Client,
}

impl RedisDistributedScheduler {
    /// 创建调度器
    pub fn new(redis_url: &str, spider_name: &str) -> Result<Self, Box<dyn Error>> {
        let client = Client::open(redis_url)?;
        let queue = RedisDistributedQueue::new(redis_url, spider_name, 1_000_000)?;
        let filter = RedisBloomFilter::new(redis_url, spider_name, 1_000_000, 0.01)?;

        Ok(Self {
            queue,
            filter,
            spider_name: spider_name.to_string(),
            client,
        })
    }

    /// 调度 URL
    pub fn schedule(&self, url: &str, priority: i32, depth: i32) -> Result<bool, Box<dyn Error>> {
        // 检查是否已爬取
        if self.filter.contains(url)? {
            return Ok(false);
        }

        // 添加到过滤器
        self.filter.add(url)?;

        // 创建任务
        let task = CrawlTask::new(url.to_string())
            .with_priority(priority)
            .with_depth(depth)
            .with_spider_name(self.spider_name.clone());

        // 入队
        self.queue.push(&task)
    }

    /// 获取下一个任务
    pub fn next_task(&self) -> Result<Option<CrawlTask>, Box<dyn Error>> {
        self.queue.pop()
    }

    /// 确认任务完成
    pub fn ack(&self, task: &CrawlTask, success: bool) -> Result<(), Box<dyn Error>> {
        self.queue.ack(task, success)?;
        self.update_stats(success)?;
        Ok(())
    }

    fn update_stats(&self, success: bool) -> Result<(), Box<dyn Error>> {
        let mut conn = self.client.get_connection()?;

        if success {
            let _: () = conn.hincr("spider:shared:stats", "success", 1)?;
            let _: () = conn.hincr("spider:shared:stats", "rust:success", 1)?;
        } else {
            let _: () = conn.hincr("spider:shared:stats", "failed", 1)?;
            let _: () = conn.hincr("spider:shared:stats", "rust:failed", 1)?;
        }
        let _: () = conn.hincr("spider:shared:stats", "processed", 1)?;
        let _: () = conn.hincr("spider:shared:stats", "rust:processed", 1)?;

        Ok(())
    }

    /// 获取统计信息
    pub fn get_stats(&self) -> Result<serde_json::Value, Box<dyn Error>> {
        let queue_stats = self.queue.get_stats()?;
        let mut conn = self.client.get_connection()?;

        let today = chrono::Local::now().format("%Y-%m-%d").to_string();
        let success: i64 = conn
            .hget(format!("stats:{}:success", self.spider_name), &today)
            .unwrap_or(0);
        let failed: i64 = conn
            .hget(format!("stats:{}:failed", self.spider_name), &today)
            .unwrap_or(0);

        Ok(serde_json::json!({
            "queue": queue_stats,
            "spider_name": self.spider_name,
            "success_today": success,
            "failed_today": failed,
            "total_processed": success + failed,
        }))
    }
}

/// Redis 工作节点
pub struct RedisWorker {
    scheduler: RedisDistributedScheduler,
    spider_name: String,
    running: bool,
    processed: usize,
    failed: usize,
    start_time: f64,
}

impl RedisWorker {
    /// 创建工作节点
    pub fn new(redis_url: &str, spider_name: &str) -> Result<Self, Box<dyn Error>> {
        let scheduler = RedisDistributedScheduler::new(redis_url, spider_name)?;

        Ok(Self {
            scheduler,
            spider_name: spider_name.to_string(),
            running: false,
            processed: 0,
            failed: 0,
            start_time: 0.0,
        })
    }

    /// 启动工作节点
    pub fn start<F>(
        &mut self,
        mut callback: F,
        max_tasks: usize,
        timeout_secs: u64,
    ) -> Result<(), Box<dyn Error>>
    where
        F: FnMut(&CrawlTask) -> Result<(), Box<dyn Error>>,
    {
        self.running = true;
        self.start_time = current_timestamp();

        println!("工作节点启动：{}", self.spider_name);

        let mut tasks_processed = 0;
        let mut last_task_time = current_timestamp();

        while self.running {
            // 检查是否达到最大任务数
            if max_tasks > 0 && tasks_processed >= max_tasks {
                println!("达到最大任务数：{}", max_tasks);
                break;
            }

            // 获取任务
            match self.scheduler.next_task()? {
                Some(task) => {
                    last_task_time = current_timestamp();
                    println!("处理任务：{}", task.url);

                    match callback(&task) {
                        Ok(()) => {
                            self.scheduler.ack(&task, true)?;
                            self.processed += 1;
                            tasks_processed += 1;
                        }
                        Err(e) => {
                            eprintln!("任务失败：{} - {}", task.url, e);
                            self.scheduler.ack(&task, false)?;
                            self.failed += 1;
                        }
                    }
                }
                None => {
                    // 队列为空，等待
                    std::thread::sleep(std::time::Duration::from_secs(1));

                    // 检查超时
                    if timeout_secs > 0
                        && (current_timestamp() - last_task_time) > timeout_secs as f64
                    {
                        println!("队列超时，退出");
                        break;
                    }
                }
            }
        }

        self.stop();
        Ok(())
    }

    /// 停止工作节点
    pub fn stop(&mut self) {
        self.running = false;
        println!("工作节点停止：{}", self.spider_name);
    }

    /// 获取统计信息
    pub fn get_stats(&self) -> serde_json::Value {
        let runtime = current_timestamp() - self.start_time;
        let rate = if runtime > 0.0 {
            self.processed as f64 / runtime
        } else {
            0.0
        };

        serde_json::json!({
            "spider_name": self.spider_name,
            "processed": self.processed,
            "failed": self.failed,
            "runtime": runtime,
            "rate": rate,
        })
    }
}

fn current_timestamp() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_bloom_filter() {
        // 注意：需要 Redis 服务
        // let filter = RedisBloomFilter::new("redis://localhost:6379", "test", 1000, 0.01).unwrap();
        // assert!(filter.add("item1").unwrap());
        // assert!(filter.contains("item1").unwrap());
    }
}

File: pyspider/distributed/redis_distributed.py
"""
Redis 分布式爬虫支持
支持分布式队列、去重、任务分发
"""

import redis
import json
import time
import hashlib
import logging
from typing import Optional, Dict, Any
from dataclasses import dataclass, asdict
from datetime import datetime
import threading

logger = logging.getLogger(__name__)


@dataclass
class CrawlTask:
    """爬取任务"""

    url: str
    priority: int = 0
    depth: int = 0
    task_type: str = "crawl"  # crawl, download, parse
    spider_name: str = "default"
    created_at: float = 0.0
    retry_count: int = 0
    metadata: Dict[str, Any] = None

    def __post_init__(self):
        if self.created_at == 0.0:
            self.created_at = time.time()
        if self.metadata is None:
            self.metadata = {}

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> "CrawlTask":
        return cls(**data)

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> "CrawlTask":
        return cls.from_dict(json.loads(json_str))


class RedisDistributedQueue:
    """Redis 分布式队列"""

    def __init__(
        self,
        name: str,
        redis_url: str = "redis://localhost:6379",
        max_size: int = 1000000,
    ):
        self.name = name
        self.max_size = max_size
        self.redis = redis.from_url(redis_url, decode_responses=True)
        self._lock = threading.RLock()

        # 统一队列 key
        self.queue_key = "spider:shared:queue"
        self.priority_queue_key = "spider:shared:queue"
        self.visited_key = "spider:shared:visited"
        self.processing_key = f"queue:{name}:processing"
        self.failed_queue_key = f"queue:{name}:failed"

    def push(self, task: CrawlTask) -> bool:
        """添加任务到队列"""
        with self._lock:
            # 检查队列大小
            if self.size() >= self.max_size:
                logger.warning(f"队列已满：{self.name}")
                return False

            # 检查是否已存在
            if self.exists(task.url):
                return False

            # 添加到优先级队列
            task_json = task.to_json()
            self.redis.zadd(self.priority_queue_key, {task_json: task.priority})

            # 添加到已访问集合 (去重)
            self.redis.sadd(self.visited_key, task.url)

            logger.debug(f"任务入队：{task.url}")
            return True

    def pop(self, timeout: float = 0) -> Optional[CrawlTask]:
        """从队列获取任务"""
        # 从优先级队列获取最高优先级的任务
        with self._lock:
            results = self.redis.zpopmin(self.priority_queue_key, count=1)

            if not results:
                return None

            task_json, score = results[0]
            task = CrawlTask.from_json(task_json)

            # 添加到处理中集合
            self.redis.hset(
                self.processing_key,
                self._url_key(task.url),
                json.dumps(
                    {"url": task.url, "started_at": time.time(), "task": task.to_dict()}
                ),
            )

            logger.debug(f"任务出队：{task.url}")
            return task

    def ack(self, task: CrawlTask, success: bool = True):
        """确认任务完成"""
        with self._lock:
            url_key = self._url_key(task.url)

            # 从处理中集合移除
            self.redis.hdel(self.processing_key, url_key)

            if not success:
                # 添加到失败队列
                if task.retry_count < 3:
                    task.retry_count += 1
                    self.push(task)
                else:
                    self.redis.lpush(self.failed_queue_key, task.to_json())

    def exists(self, url: str) -> bool:
        """检查 URL 是否已访问"""
        return self.redis.sismember(self.visited_key, url)

    def size(self) -> int:
        """获取队列大小"""
        return self.redis.zcard(self.priority_queue_key)

    def processing_count(self) -> int:
        """获取处理中的任务数"""
        return self.redis.hlen(self.processing_key)

    def failed_count(self) -> int:
        """获取失败队列大小"""
        return self.redis.llen(self.failed_queue_key)

    def clear(self):
        """清空队列"""
        with self._lock:
            self.redis.delete(
                self.queue_key,
                f"{self.queue_key}:urls",
                self.priority_queue_key,
                self.processing_key,
                self.failed_queue_key,
            )

    def _url_key(self, url: str) -> str:
        """生成 URL 键"""
        return hashlib.md5(url.encode()).hexdigest()

    def get_stats(self) -> dict:
        """获取统计信息"""
        return {
            "queue_size": self.size(),
            "processing": self.processing_count(),
            "failed": self.failed_count(),
            "max_size": self.max_size,
        }


class RedisBloomFilter:
    """Redis 布隆过滤器（分布式去重）"""

    def __init__(
        self,
        name: str,
        redis_url: str = "redis://localhost:6379",
        expected_items: int = 1000000,
        error_rate: float = 0.01,
    ):
        self.name = name
        self.redis = redis.from_url(redis_url, decode_responses=True)
        self.key = f"bloom:{name}"

        # 计算布隆过滤器参数
        import math

        self.size = int(-expected_items * math.log(error_rate) / (math.log(2) ** 2))
        self.hash_count = int((self.size / expected_items) * math.log(2))

        # 调整大小
        self.size = ((self.size + 7) // 8) * 8  # 对齐到字节

    def add(self, item: str) -> bool:
        """添加元素，返回是否可能已存在"""
        item_bytes = item.encode("utf-8")
        is_new = True

        for seed in range(self.hash_count):
            pos = self._hash(item_bytes, seed) % self.size
            if not self.redis.setbit(self.key, pos, 1):
                is_new = False

        return is_new

    def contains(self, item: str) -> bool:
        """检查元素是否存在"""
        item_bytes = item.encode("utf-8")

        for seed in range(self.hash_count):
            pos = self._hash(item_bytes, seed) % self.size
            if not self.redis.getbit(self.key, pos):
                return False

        return True

    def _hash(self, item: bytes, seed: int) -> int:
        """哈希函数"""
        return int(hashlib.md5(item + str(seed).encode()).hexdigest(), 16)

    def clear(self):
        """清空过滤器"""
        self.redis.delete(self.key)

    def count(self) -> int:
        """估算元素数量"""
        import math

        bits_set = self.redis.bitcount(self.key)
        if bits_set == 0:
            return 0
        return int(-self.size / self.hash_count * math.log(1 - bits_set / self.size))


class RedisDistributedScheduler:
    """Redis 分布式调度器"""

    def __init__(
        self,
        spider_name: str,
        redis_url: str = "redis://localhost:6379",
        queue_cls=RedisDistributedQueue,
    ):
        self.spider_name = spider_name
        self.redis = redis.from_url(redis_url, decode_responses=True)

        # 任务队列
        self.queue = queue_cls(spider_name, redis_url)

        # 布隆过滤器（去重）
        self.filter = RedisBloomFilter(spider_name, redis_url)

        # 统计信息
        self.stats_key = "spider:shared:stats"

    def schedule(self, url: str, priority: int = 0, depth: int = 0, **metadata) -> bool:
        """调度 URL"""
        # 检查是否已爬取
        if self.filter.contains(url):
            logger.debug(f"URL 已存在：{url}")
            return False

        # 添加到过滤器
        self.filter.add(url)

        # 创建任务
        task = CrawlTask(
            url=url,
            priority=priority,
            depth=depth,
            spider_name=self.spider_name,
            metadata=metadata,
        )

        # 入队
        return self.queue.push(task)

    def next_task(self) -> Optional[CrawlTask]:
        """获取下一个任务"""
        return self.queue.pop()

    def ack(self, task: CrawlTask, success: bool = True):
        """确认任务完成"""
        self.queue.ack(task, success)
        self._update_stats(success)

    def _update_stats(self, success: bool):
        """更新统计"""
        if success:
            self.redis.hincrby(self.stats_key, "success", 1)
            self.redis.hincrby(self.stats_key, "python:success", 1)
        else:
            self.redis.hincrby(self.stats_key, "failed", 1)
            self.redis.hincrby(self.stats_key, "python:failed", 1)

        self.redis.hincrby(self.stats_key, "processed", 1)
        self.redis.hincrby(self.stats_key, "python:processed", 1)

    def get_stats(self) -> dict:
        """获取统计信息"""
        queue_stats = self.queue.get_stats()

        # 获取历史统计
        today = datetime.now().strftime("%Y-%m-%d")
        success = int(self.redis.hget(f"{self.stats_key}:success", today) or 0)
        failed = int(self.redis.hget(f"{self.stats_key}:failed", today) or 0)

        return {
            **queue_stats,
            "spider_name": self.spider_name,
            "success_today": success,
            "failed_today": failed,
            "total_processed": success + failed,
        }

    def clear(self):
        """清空调度器"""
        self.queue.clear()
        self.filter.clear()


class RedisWorker:
    """Redis 分布式工作节点"""

    def __init__(
        self,
        spider_name: str,
        redis_url: str = "redis://localhost:6379",
        callback: callable = None,
    ):
        self.spider_name = spider_name
        self.redis = redis.from_url(redis_url, decode_responses=True)
        self.callback = callback

        self.scheduler = RedisDistributedScheduler(spider_name, redis_url)
        self.running = False
        self.stats = {
            "processed": 0,
            "failed": 0,
            "start_time": None,
            "last_task_time": None,
        }

    def start(self, max_tasks: int = 0, timeout: int = 300):
        """启动工作节点"""
        self.running = True
        self.stats["start_time"] = time.time()

        logger.info(f"工作节点启动：{self.spider_name}")

        tasks_processed = 0

        while self.running:
            # 检查是否达到最大任务数
            if max_tasks > 0 and tasks_processed >= max_tasks:
                logger.info(f"达到最大任务数：{max_tasks}")
                break

            # 获取任务
            task = self.scheduler.next_task()

            if task:
                self.stats["last_task_time"] = time.time()
                logger.info(f"处理任务：{task.url}")

                try:
                    # 执行回调
                    if self.callback:
                        self.callback(task)

                    # 确认成功
                    self.scheduler.ack(task, success=True)
                    self.stats["processed"] += 1
                    tasks_processed += 1

                except Exception as e:
                    logger.error(f"任务失败：{task.url} - {e}")
                    self.scheduler.ack(task, success=False)
                    self.stats["failed"] += 1
            else:
                # 队列为空，等待
                time.sleep(1)

                # 检查超时
                if (
                    timeout > 0
                    and (time.time() - self.stats["last_task_time"]) > timeout
                ):
                    logger.info("队列超时，退出")
                    break

        self.stop()

    def stop(self):
        """停止工作节点"""
        self.running = False
        logger.info(f"工作节点停止：{self.spider_name}")

    def get_stats(self) -> dict:
        """获取统计信息"""
        scheduler_stats = self.scheduler.get_stats()

        runtime = (
            time.time() - self.stats["start_time"] if self.stats["start_time"] else 0
        )
        rate = self.stats["processed"] / runtime if runtime > 0 else 0

        return {
            **scheduler_stats,
            "processed": self.stats["processed"],
            "failed": self.stats["failed"],
            "runtime": runtime,
            "rate": rate,
        }


# 使用示例
if __name__ == "__main__":
    # 创建调度器
    scheduler = RedisDistributedScheduler(
        spider_name="test_spider", redis_url="redis://localhost:6379"
    )

    # 添加任务
    scheduler.schedule("https://example.com/1", priority=10)
    scheduler.schedule("https://example.com/2", priority=5)
    scheduler.schedule("https://example.com/3", priority=15)

    print(f"队列大小：{scheduler.queue.size()}")
    print(f"统计：{scheduler.get_stats()}")

    # 创建工作节点
    def process_task(task: CrawlTask):
        print(f"处理：{task.url}")
        time.sleep(1)  # 模拟处理

    worker = RedisWorker(
        spider_name="test_spider",
        redis_url="redis://localhost:6379",
        callback=process_task,
    )

    # 启动 worker
    worker.start(max_tasks=10)

    print(f"最终统计：{worker.get_stats()}")
